samples/coop_synchronization/src/main.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
+ * Copyright (c) 2023 Advanced Micro Devices, Inc. (AMD)
+ * Copyright (c) 2023 Alp Sayin <alpsayin@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+
+/*
+ * The coop synchronization demo has two threads that utilize semaphores and k_yield
+ * to take turns printing a greeting message with a somewhat uncontrolled rate.
+ * The rate is controlled via a busy waiting loop. The advantage of this demo is that
+ * it'll work even if there are no timing utilities provided by the architecture.
+ */
+
+#define PIN_THREADS (IS_ENABLED(CONFIG_SMP)		  \
+		     && IS_ENABLED(CONFIG_SCHED_CPU_MASK) \
+		     && (CONFIG_MP_NUM_CPUS > 1))
+
+/* size of stack area used by each thread */
+#define STACKSIZE 1024
+
+/* scheduling priority used by each thread */
+#define PRIORITY 7
+
+/*
+ * @param my_name      thread identification string
+ * @param my_sem       thread's own semaphore
+ * @param other_sem    other thread's semaphore
+ */
+void helloLoop(const char *my_name, struct k_sem *my_sem,
+	struct k_sem *other_sem, int incrementer)
+{
+	const char *tname;
+	uint8_t cpu;
+	struct k_thread *current_thread;
+	int task_local_counter = 0;
+
+	while (1) {
+		/* take my semaphore */
+		k_sem_take(my_sem, K_FOREVER);
+
+		int key1 = irq_lock();
+		{
+			current_thread = k_current_get();
+			tname = k_thread_name_get(current_thread);
+#if CONFIG_SMP
+			cpu = arch_curr_cpu()->id;
+#else
+			cpu = 0;
+#endif
+			printk("%s: level 1 irq_lock key = %d\n", tname, key1);
+			int key2 = irq_lock();
+			{
+				k_yield();
+				printk("%s: level 2 irq_lock key = %d\n", tname, key2);
+			}
+			irq_unlock(key2);
+		}
+		irq_unlock(key1);
+		/* say "hello" */
+		if (tname == NULL) {
+			printk("%s: Hello World from cpu %d on %s! counter = %d\n", my_name, cpu,
+						 CONFIG_BOARD, task_local_counter += incrementer);
+		} else {
+			printk("%s: Hello World from cpu %d on %s! counter = %d\n", tname, cpu,
+						 CONFIG_BOARD, task_local_counter += incrementer);
+		}
+
+		/* wait a while, then let other thread have a turn */
+		for (int i = 0; i < 100000000; i++) {
+			arch_nop();
```

Comment:
I had a comment here telling you to use k_busy_wait() instead of rolling your own idle loop, but then realized that's timer-based and so won't work in the domain you want to exercise.  That subtlety is worth a comment here.

Also note that bespoke busy loops don't play well with native_posix, which can't preempt threads except at specific points.  There is a brand new Z_SPIN_DELAY() API intended for situations like this that will reschedule as needed.  As it happens, you don't need that either (since nothing can preempt you), but the fact that you don't is surprising and, again, probably worth explaining.

Reply:
> As it happens, you don't need that either (since nothing can preempt you), but the fact that you don't is surprising and, again, probably worth explaining.
😄 Done. Thank you very much. Indeed, some things I've done during development has become "natural" to me that I hadn't drop notes.